### ISSUE TITLE
Newman behind corporate proxy fails with "socket hang up"

### ISSUE THREAD

**Issue by cbarrerah-sogeti** [postmanlabs/newman#818]
1. Newman Version (can be found via `newman -v`): 3.3.0
 2. OS details (type, version, and architecture): Windows7 Enterprise, 64 Bit Intel Core i7
 3. Are you using Newman as a library, or via the CLI? CLI (through Jenkins and direct bash script)
 3. Did you encounter this recently, or has this bug always been there: Only using newman since 3.3.0
 4. Expected behaviour: Consistent results for the testing done to the collection, with no random failures due to proxy
 5. Command / script used to run Newman: newman run Collection.json -e env.json -k --disable-unicode -r junit,cli
 6. Sample collection, and auxilliary files (minus the sensitive details):
I made a test collection pointing to Httpbin.org

{
	"variables": [],
	"info": {
		"name": "HTTPBin",
		"_postman_id": "ba9d25c1-bcd2-4a7e-cd8c-baa37b4af407",
		"description": "Base API testing collection to serve as a positive control for Jenkins\n",
		"schema": "https://schema.getpostman.com/json/collection/v2.0.0/collection.json"
	},
	"item": [
		{
			"name": "get",
			"event": [
				{
					"listen": "test",
					"script": {
						"type": "text/javascript",
						"exec": [
							"tests[\"Status code is 200\"] = responseCode.code === 200;",
							"",
							"var jsonData = JSON.parse(responseBody);",
							"tests[\"Host correct\"] = jsonData.headers.Host == \"httpbin.org\";"
						]
					}
				}
			],
			"request": {
				"url": "https://httpbin.org/get",
				"method": "GET",
				"header": [],
				"body": {},
				"description": ""
			},
			"response": []
		}
	]
}

and the env.json file:
{
  "id": "35c0b2b8-6506-e43b-f39f-242923416836",
  "name": "HTTPBin",
  "values": [
    {
      "key": "HTTPS_PROXY",
      "type": "text",
      "value": "http://proxyURL:8080",
      "enabled": true
    },
    {
      "key": "HTTP_PROXY",
      "type": "text",
      "value": "http://proxyURL:8080",
      "enabled": true
    }
  ],
  "timestamp": 1481796886584,
  "_postman_variable_scope": "environment",
  "_postman_exported_at": "2016-12-15T10:14:49.984Z",
  "_postman_exported_using": "Postman/4.9.2"
}
The test.sh script called from Jenkins is:
#!/usr/bin/bash
/path/to/newman run Collection.json -e env.json -k --disable-unicode -r junit,cli
 7. Screenshots (if applicable):
Fail in Jenkins
![image](https://cloud.githubusercontent.com/assets/24520879/21220195/147ea8d4-c2b8-11e6-8712-c9dd1b0b5b73.png)
Success in Jenkins
![image](https://cloud.githubusercontent.com/assets/24520879/21220212/278e50aa-c2b8-11e6-9740-b735b470923e.png)

Fail in CLI
![image](https://cloud.githubusercontent.com/assets/24520879/21220316/97372b70-c2b8-11e6-8b06-fd4cd71cd1ee.png)
Success in CLI
![image](https://cloud.githubusercontent.com/assets/24520879/21220333/ac6db68a-c2b8-11e6-88ea-8187a3448df7.png)

Steps to reproduce the problem:

1. Execute the script in the same folder the env.json and Collection.json files
2. Repeat 10-15 times every 2-5 miuntes
3. Alternatively, set up a Jenkins job and analyze the failure pattern
![image](https://cloud.githubusercontent.com/assets/24520879/21220403/fbd28660-c2b8-11e6-88ae-d56ded7293f4.png)

The only sensitive information is the corporate proxy setup. In the internet settings panel, is set up with a setup script:
![image](https://cloud.githubusercontent.com/assets/24520879/21220497/67833f76-c2b9-11e6-90bb-290a87210ec6.png)

I located a normal proxy inside the file and use it as the proxy in Postman and set it up with HTTP_PROXY/HTTPS_PROXY environment variables for the newman run.

The ideal solution would be for Postman/newman to be able to digest proxy configuration scripts directly through the url, but handling the proxy correctly by direct input would be enough for me.

Thanks in advance and if you need any additional information, i'd be glad to help.

**Comment by czardoz** [comment]
@cbarrerah-sogeti That's your proxy dropping the connection from Newman, and there isn't anything we can do about that. You can try adding a "User-Agent" header set to the user agent of some browser, and see if it fools the proxy into letting the connections through :-)

**Comment by cbarrerah-sogeti** [comment]
Implemented a rotation of User-Agent for the executions with a pre-request script:
var counter = Number(postman.getEnvironmentVariable("userAgentCounter"));
//console.log(counter);
if (counter >= 5) {
    // Let's rotate the User-Agent
    userIndex = Number(postman.getEnvironmentVariable("userAgentIndex"));
    userIndex = (userIndex+1)%4;
    console.log(userIndex);
    postman.setEnvironmentVariable("userAgentIndex", userIndex);
    var userAgents = ["Mozilla/5.0 (Windows NT 6.1; Win64; x64)", "AppleWebKit/537.36 (KHTML, like Gecko)", "Chrome/55.0.2883.87",  "Safari/537.36"];
    var agent = userAgents[userIndex];
    console.log(agent);
    postman.setEnvironmentVariable("userAgent", agent);
    counter = 0; 
    //postman.setEnvironmentVariable("userAgentCounter", 0)
}
counter = counter + 1 ;
postman.setEnvironmentVariable("userAgentCounter", counter);

Same rate of success / failure :(

Any hot nail to grab for me? :)

**Comment by cbarrerah-sogeti** [comment]
More interesting still, the same collection with the same envVars, but run from postman, behaves as expected and can be run multiple times even with tenths of iterations at the same time.

So it seems to be a newman specific issue. Maybe newman is not handling the HTTP_PROXY environment variables correctly?

**Comment by czardoz** [comment]
@cbarrerah-sogeti Postman uses the APIs provided by the OS to detect the Proxy settings. As a CLI, Newman does not do that, which explains the difference you see :)

**Comment by cbarrerah-sogeti** [comment]
and is there any way to let newman know how to use the proxy?
I had a similar problem with npm not updating or installing anything because of the proxy, but i set up some env vars for http-proxy and https-proxy, and everything became responsive again. Is there any other mechanism for newman that may work or is there no way for newman to work behind a proxy?

Thanks in advance and merry christmas ;)

**Comment by czardoz** [comment]
@cbarrerah-sogeti yes, similar to NPM, you can set the `HTTP_PROXY` and `HTTPS_PROXY` environment variables, and Newman will automatically start using them :)

**Comment by cbarrerah-sogeti** [comment]
that's the weird spot, i have set the environment variables HTTP_PROXY and HTTPS_PROXY in the collection run, and newman seems to be ignoring them.

**Comment by perrefe** [comment]
Hi, I'm getting the same situation but I'm not using a corporate proxy. In my case, an haproxy is forwarding requests to the endpoint.

Through haproxy, some newman requests fail because of a "socket hang up at request". Doing the same but requesting direct to the endpoint's ip:port, tests are passing.

**Comment by czardoz** [comment]
@perrefe Probably an issue with your HAProxy configuration then?

**Comment by perrefe** [comment]
Probably not, because the real client application is requesting through the same haproxy and works well.

**Comment by czardoz** [comment]
@perrefe Hmm, can you open a new issue for this with more details? I doubt it is a Newman issue because you mentioned that it works fine when you directly call your endpoint (bypassing HAProxy)

**Comment by cbarrerah-sogeti** [comment]
@perrefe if you open a new issue for this, please add the link to this thread as i will follow it care.

@czardoz i think what @perrefe said that the real client app is requesting through the same proxy and works well, but when the newman run is passed through the proxy fails, while going directly bypassing the proxy, it works again.

By this description i assume that the problem is in the way that newman handles the proxied connection, not the proxy itself.
That is consistent with the issue i had, because the same tests run manually with Postman work well, even with the collection runner, but when the collection is passed to newman, the same tests get a "socket hang up at request error".

So the summary of both cases is:
- Collection run through proxy with postman, tests work as expected
- Collection run through proxy with newman, test get a "socket hang up at request error"

**Comment by czardoz** [comment]
@cbarrerah-sogeti Actually, @perrefe's issue is different from yours, because HAProxy is a _reverse_ proxy, i.e, when it's being used, Newman does not need to be specifically configured for it.

In your case, it's a _forward_ proxy configuration issue. I'm re-opening this, since you're still facing a problem.

**Comment by cbarrerah-sogeti** [comment]
Hi, i have news regarding this issue.

I tried configurind the environment variables for Http_proxy/https_proxy not in the environment variables exported from postman, but in the environment variables configured in the jenkins setup itself.

And now it works!

What is troubling is that the environment variables that i'm explicitly sending through the environment.json file exported from Postman is not "convincing" enough for newman to use it.

Anyways, if naybody else stumbles upon the same behaviour, the go to point is:
Jenkins->Manage Jenkins->Configure System->Global Properties/Environment Variables
I added the uppercase HTTP_PROXY/HTTPS_PROXY combination and it started working.

Happy days!
